api/app/auto_spatial_advisory/elevation.py

```diff
@@ -178,6 +210,63 @@ def apply_threshold_mask_to_dem(threshold: int, mask_path: str, temp_dir: str):
     return masked_dem_path
 
 
+async def process_tpi_by_firezone(run_type: RunType, run_date: date, for_date: date) -> Dict[int, Dict[int, int]]:
+    """
+    Given run parameters, lookup associated snow-masked HFI and static classified TPI geospatial data.
+    Cut out each fire zone shape from the above and intersect the TPI and HFI pixels, counting each pixel contributing to the TPI class.
+    Capture all fire zone stats keyed by its source_identifier.
+
+    :param run_parameters_id: The RunParameter object id associated with this run_type, for_date and run_datetime
+    :return: dictionary of {source_identifier: {1: X, 2: Y, 3: Z}}, where 1 = valley bottom, 2 = mid slope, 3 = upper slope
+    and X, Y, Z are pixel counts at each of those elevation classes respectively.
+    """
+    raster_options = GeospatialOptions(include_geotransform=True, include_projection=True, include_x_size=True, include_y_size=True)
+    fire_zone_stats: Dict[int, Dict[int, int]] = {}
+
+    hfi_raster_filename = get_raster_tif_filename(for_date)
+    hfi_raster_key = get_raster_filepath(run_date, run_type, hfi_raster_filename)
+
+    async with get_client() as (client, bucket):
+        tpi_result = await read_raster_into_memory(
+            client,
+            bucket,
+            f'dem/tpi/{config.get("CLASSIFIED_TPI_DEM_NAME")}',
+            raster_options,
+        )
+        hfi_result = await read_raster_into_memory(
+            client,
+            bucket,
+            hfi_raster_key,
+            GeospatialOptions(
+                warp_options=WarpOptions(
+                    source_geotransform=tpi_result.data_geotransform,
+                    source_projection=tpi_result.data_projection,
+                    source_x_size=tpi_result.data_x_size,
+                    source_y_size=tpi_result.data_y_size,
+                )
+            ),
+        )
+
+        async with get_async_write_session_scope() as session:
+            stmt = text("SELECT id, source_identifier FROM advisory_shapes;")
+            result = await session.execute(stmt)
+            rows = result.all()
+
+            for row in rows:
+                cut_tpi_result = cut_raster_by_shape_id(row[0], row[1], tpi_result.data_source, raster_options)
```

Comment:
Not a big deal, but might be slightly simpler computationally to multiply the two rasters first outside of the for loop, then cut out the advisory shapes from the result.

Reply:
Good call, thanks, I'll make that change.